src/flb_io.c

```diff
@@ -307,12 +313,67 @@ static ssize_t net_io_read(struct flb_upstream_conn *u_conn,
                            void *buf, size_t len)
 {
     int ret;
+    struct pollfd pfd_read;
+
+    /* Set socket to non-blocking mode for timeout */
+    flb_net_socket_nonblocking(u_conn->fd);
```

Comment:
If the socket is going to be set to nonblocking on connect as well as every read and write, why not make all sockets nonblocking and avoid creating blocking sockets? Where would we be using sockets in blocking mode after this change?

Reply:
Is setting a receive timeout (`SO_RCVTIMEO`) and the equivalent send timeout on the blocking socket not an option or not portable enough? Does the timeout need to apply to calls other than recv and send?

Reply:
I don't think there should be any issues with using `SO_RCVTIMEO` and `SO_SNDTIMEO`. Additionally, I'm not really sure of this juggling with the socket state thing, I don't really like it.